Add extract functionality as a new criteria to timestamptz columns

jinn999 opened the pull request:
This PR concerns the extract info from date mentioned in issue #857.

It provides a way to be able to extract specific fields from a `timestamptz` column and compare it with some numerical values.
For all the fields implemented till now (see `Avram::ChronoUnits` in the `extract_criteria.cr` or the relevant postgresql documentation), three possible methods are provided:

- `column.extract(Avram::ChronoUnits::<enum_value>)`
             e.g: `some_table.created_on.extract(Avram::ChronoUnits::Day)`
- `column.extract(<enum_value_as_symbol>)`
             e.g: `some_table.created_on.extract(:day)`
- `column.extract_<enum_value>`
             e.g: `some_table.created_on.extract_day`

Functionality can be used as:
- column.extract_year.eq(2022)

Most values returned are ints, but 4 of them which return floats
- julian
- epoch
- second
- milliseconds

I am still not able to do the whole script/setup script/test stuff due to the M1 usual issue, but relevant specs pass..

jwoertink commented on src/avram/criteria_extensions/extract_criteria.cr:
```diff
@@ -0,0 +1,132 @@
+    Quarter
+    Second
+    Timezone
+    Timezone_hour
+    Timezone_minute
```
Curious about these. Is this because of the Symbol arg override? I'm assuming doing TimezoneHour then passing :timezone_hour fails?

jinn999 replied:
Actually, it was a complete mishap from my part, dunno why but I was convinced that they were with no underscore :/ either way, I changed them to be correct now :) which also made me learn about camelcase and underscore methods XD

jwoertink commented on src/avram/criteria_extensions/extract_criteria.cr:
```diff
@@ -0,0 +1,132 @@
+    def extract_year
+        extract(Avram::ChronoUnits::Year)
+    end
+
+end
+end
```
I can't tell if this is formatting or the GitHub mobile app 😅

jinn999 replied:
Now, this is funny. it was not the mobile app, it was formatting, but the funny thing is that neither tool format nor ameba had anything to complain :/

jwoertink replied:
That's strange. You'd think the formatter would have caught that 🤔

jinn999 replied:
Yup, pretty strange... But also the checks you have in the pipeline would have caught that :/ bug in the crystal tool format? Dunno :/

jwoertink reviewed (commented):
This is amazing! Thanks so much for adding this in. A few small comments but I think it's great 😁

jinn999 commented:
Ok, did a major revamp to avoid having all these extract_xxx methods to define and risking forgetting some of them. Switched to heavy macroing, and moved chronounits in their own file

jwoertink reviewed (approved):
:+1:  Thanks!

paulcsmith commented:
This is super cool. Thanks @jinn999
